opentelemetry/api/trace/__init__.py

```diff
@@ -0,0 +1,190 @@
+    with tracer.span("parent"):
+        # Attach a new child and update the current span
+        with tracer.span("child"):
+            do_work():
+        # Close child span, set parent as current
+    # Close parent span, set default span as current
+
+Under explicit context propagation there is no concept of an active span, and
+the caller is responsible for managing the span's lifetime::
+
+    from opentelemetry.sdk.trace import tracer
+    from your.integration import deserialize_span
+
+    parent = deserialize_span(serialized_span)
+    # Explicit parent span assignment
+    with tracer.span("child", parent=parent) as child:
+        do_work(span=child)
+
+Applications should generally use a single global tracer, and use either
+implicit or explicit context propagation consistently throughout.
+
+.. versionadded:: 0.1.0
+"""
+
+from __future__ import annotations
+
+from typing import Iterator
+from contextlib import contextmanager
+
+
+class Tracer(object):
+    """Handles span creation and in-process context propagation.
+
+    This class provides methods for manipulating the context, creating spans,
+    and controlling spans' lifecycles.
+    """
+
+    def get_current_span(self) -> Span:
+        """Get the currently active span from the context.
+
+        Returns:
+            The currently active :class:`.Span`.
+        """
+        raise NotImplementedError
+
+    @contextmanager
+    def span(self, name: str, parent: Span=None) -> Iterator[None]:
+        """Context manager for span creation.
+
+        Create a new child of the current span, or create a root span if no
+        current span exists. Start the span and set it as the current span in
+        this tracer's context.
+
+        On exiting the context manager stop the span and set its parent as the
+        current span.
+
+        Example::
+
+            with tracer.span("one") as parent:
+                parent.add_event("parent's event")
+                with tracer.span("two") as child:
+                    child.add_event("child's event")
+                    tracer.get_current_span()  # returns child
+                tracer.get_current_span()      # returns parent
+            tracer.get_current_span()          # returns the previously active span
+
+        Args:
+            name: The name of the span to be created.
+            parent: This span's parent.
+
+        Raises:
+            ValueError: if ``name`` is null.
+        """
+        raise NotImplementedError
+
+    def create_span(self, name: str, parent: Span=None) -> Span:
+        """Create a new span as a child of the currently active span.
+
+        If called with ``parent`` this method won't affect the tracer's
+        context.
+
+        See ``span`` for a context manager that controls the span's lifetime.
+
+        Args:
+            name: The name of the span to be created.
+            parent: This span's parent.
+
+        Raises:
+            ValueError: if ``name`` is null.
```

Comment:
Do we only check if `name` is null or we expect more validations?

Reply:
I'd personally advocate for using a generic name here, instead of failing.

Reply:
OK, it sounds like implementations should provide a safe default here. In general, do you expect implementations to be more or less permissive than the API? I.e. should implementations be required to throw errors we list in the API? Should they be allowed to throw errors we don't list?

Reply:
> In general, do you expect implementations to be more or less permissive than the API? I.e. should implementations be required to throw errors we list in the API?

Yes, I bet almost all implementations will have more restriction than the API would allow.

> Should they be allowed to throw errors we don't list?

Probably not. One design principle we may want to follow is "the monitoring SDK will try to keep silent instead of changing the application behavior".  In this spirit I guess we will have exporters doing some downcast/trimming based on their specific need, and not throw exceptions.

Reply:
In that case: do we want to allow implementations to throw here on null/blank `name` by including it in the API or are we ok with forbidding them from throwing by removing it?

Reply:
I could think of the following approaches:
1. put something like "<unnamed>" or "N/A" as the span name if we got null/blank
2. throw `ValueError`
3. generate a random span name in case we got null/blank

Personally I'm in favor of option 1.

Reply:
> I bet almost all implementations will have more restriction than the API would allow

This suggests that implementations will throw errors we don't list in the API, or am I missing something?

FWIW the java API says to [raise a NPE here](https://github.com/c24t/opentelemetry-java/blob/8b1865303ae46ff419a7dc803fcd4e2552b45314/api/src/main/java/io/opentelemetry/trace/Tracer.java#L160). It might be worth taking this conversation there (or to the spec repo).

Using a safe default span name sounds good to me as long as we get the same behavior in all clients, or decide that this is "unspecified behavior" and that different clients can do this differently.

Reply:
> Personally I'm in favor of option 1.

I'm also in favor one. That being said, this can also be mentioned/discussed in the Specification ;)

Reply:
I'd also prefer option 1. Option 3 (generating a random span name) is no good IMHO, because span names should be low-cardinality.

Reply:
I added https://github.com/open-telemetry/opentelemetry-specification/issues/141 to track this in specs.

Reply:
It sounds like we should list any error implementations _could_ throw in the api method docstring. Safe default names sound good to me, but let's clarify this in the spec first. I added https://github.com/open-telemetry/opentelemetry-python/issues/18 to track this.